pageserver/src/layered_repository/timeline.rs

```diff
@@ -837,92 +930,153 @@ impl Timeline {
+                                                                .initial_logical_size
+                                                                .set(calculated_size) {
+                        error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing")
+                    };
+
+                    finish_sender.send(()).ok();
+                    Ok(())
+                }) {
+                Ok(guard) => *task_guard = Some((guard, finish_receiver)),
+                Err(e) => error!("Failed to spawn timeline {timeline_id} size init task: {e}"),
+            }
+        }
+    }
+
+    /// Calculate the logical size of the database at the latest LSN.
+    ///
+    /// NOTE: counted incrementally, includes ancestors, this can be a slow operation.
+    fn calculate_logical_size(&self, up_to_lsn: Lsn) -> anyhow::Result<u64> {
+        info!("Calculating logical size for timeline {}", self.timeline_id);
+
         // Try a fast-path first:
         // Copy logical size from ancestor timeline if there has been no changes on this
         // branch, and no changes on the ancestor branch since the branch point.
-        if self.get_ancestor_lsn() == self.get_last_record_lsn() && self.ancestor_timeline.is_some()
-        {
+        let ancestor_branch_off_lsn = self.get_ancestor_lsn();
+        if self.ancestor_timeline.is_some() && up_to_lsn == ancestor_branch_off_lsn {
             let ancestor = self.get_ancestor_timeline()?;
-            let ancestor_logical_size = ancestor.get_current_logical_size();
-            // Check LSN after getting logical size to exclude race condition
-            // when ancestor timeline is concurrently updated.
-            //
-            // Logical size 0 means that it was not initialized, so don't believe that.
-            if ancestor_logical_size != 0 && ancestor.get_last_record_lsn() == self.ancestor_lsn {
-                self.set_current_logical_size(ancestor_logical_size);
-                debug!(
-                    "logical size copied from ancestor: {}",
-                    ancestor_logical_size
-                );
-                return Ok(());
+            if ancestor.get_last_record_lsn() == ancestor_branch_off_lsn {
+                let ancestor_logical_size = ancestor
+                    // Note, that while we calculate some logical size based on the ancestor layers only,
+                    // we don't save this result to ancestor's data: ancestor could have its `initial_part_end`
+                    // not equal to `up_to_lsn`/`ancestor_branch_off_lsn`.
+                    //
+                    // We also don't have a function that accepts `from_lsn: Lsn` to calculate a logical size,
+                    // limited with the lower value.
+                    //
+                    // Due to these limitations, we never reuse other logical size calculation results and recompute
```

Comment:
I've rewritten the code so that the size init is done only once, as all the comments before and the naming implied.
Due to such approach, we traverse ancestors and record their sizes for init once only.

Before, we used the following code to do that:

https://github.com/neondatabase/neon/blob/1a666a01d672298a1da12771c4b68a28c1d5ebed/pageserver/src/layered_repository/timeline.rs#L854

and this way, we received incomplete ancestors' sizes, hence potentially reporting less timeline size as it is in reality and never updated that value with ancestors' changes.

Now, we wait for all changes to arrive, to be precise, yet that fails our `debug` tests: we're unable to calculate some sizes for half a minute at least, see the corresponding failures.

I'm not quite sure what to do now, except changing the test and adding some TODO, since this looks like a big separate task.